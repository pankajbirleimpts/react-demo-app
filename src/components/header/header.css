
.header-container {
  margin-top: 15px !important;
}
