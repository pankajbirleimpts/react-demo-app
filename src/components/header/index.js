import React from "react";
import { connect } from "react-redux";
import { Link, useRouteMatch, withRouter } from "react-router-dom";
import { Grid, Image } from "semantic-ui-react";
import "./header.css";
import { signout } from "../../actions/UserAction";

export function Header({ user, signout, history }) {
  const signoutToUser = () => {
    signout(() => {
      history.push("/login");
    });
  };

  return (
    <Grid className="header-container">
      <Grid.Row className="header-name-section">
        <Grid.Column width={12}>
          <h3 className="web-logo">Canteen Management System</h3>
        </Grid.Column>
        {user.isAuthenticated === true && (
          <Grid.Column width={4} data-test="user-details-section">
            <div className="user-details">
              <p className="user-name">
                {user.data.firstName} {user.data.lastName}
              </p>
              <p className="user-role">({user.data.role})</p>
              <a href="javasctipt:void(0)" data-test="logout-action" onClick={() => signoutToUser()}>
                Logout
              </a>
            </div>
          </Grid.Column>
        )}
      </Grid.Row>

      <Grid.Row className="menu-container">
        <Grid.Column>
          <div className="ui secondary pointing menu">
            {user.isAuthenticated === true && (
              <React.Fragment key="user-menu">
                <MenuLink to="/dashboard" label="Dashboard" />
                {user.data && user.data.role && user.data.role === "ADMIN" && (
                  <React.Fragment key="admin-nav">
                    <MenuLink data-test="admin-navigation" to="/users" label="Users" />
                    <MenuLink to="/items" label="Items" />
                    <MenuLink to="/day-items" label="Day Items" />
                    <MenuLink to="/purchase-item" label="Purchase Item" />
                  </React.Fragment>
                )}
                <MenuLink to="/transactions" label="Transactions" />
              </React.Fragment>
            )}
            {user.isAuthenticated !== true && (
              <React.Fragment >
                <MenuLink  to="/" label="Home" />
                <MenuLink data-test="sign-navigation" to="/signup" label="Signup" />
              </React.Fragment>
            )}
          </div>
        </Grid.Column>
      </Grid.Row>
    </Grid>
  );
}

function MenuLink({ label, to, activeOnlyWhenExact = true }) {
  let match = useRouteMatch({
    path: to,
    exact: activeOnlyWhenExact
  });

  return (
    <Link className={match ? "item active" : "item"} to={to}>
      {label}
    </Link>
  );
}

const mapStateToProps = state => ({
  user: state.user
});

export default connect(mapStateToProps, { signout })(withRouter(Header));
