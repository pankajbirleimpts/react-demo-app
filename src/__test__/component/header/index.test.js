import React from 'react';
import { shallow, mount } from 'enzyme';
import { Header } from '../../../components/header/index';
import { storeFactory, findByAttr } from '../../test/TestUtils';

const setup = (props) => {
    const wrapper = shallow(
        <Header {...props} />,
    )
    return wrapper;
};

describe('If a user is login with admin', () => {
    const props = {
        user: {
            isAuthenticated: true,
            data: {
                firstName: 'Admin',
                lastName: 'User',
                role: 'ADMIN',
            }
        },
        signout: jest.fn(),
        history: {
            push: jest.fn()
        }
    };

    test('Snapshot of Header component', () => {
        const wrapper = setup(props);
        expect(wrapper).toMatchSnapshot();
    });


    test('Check user details section should be displayed', () => {
        const wrapper = setup(props);
        const userDetailsSection = findByAttr(wrapper, 'user-details-section');
        expect(userDetailsSection.length).toBe(1);
    });

    test('Admin navaigation section should be displayed', () => {
        const wrapper = setup(props);
        const adminNavigationSection = findByAttr(wrapper, 'admin-navigation');
        expect(adminNavigationSection.length).toBe(1);
    });

    test('After clicking on signout, `signout` action should be called', () => {
        const wrapper = setup(props);
        const logoutAction = findByAttr(wrapper, 'logout-action');
        logoutAction.simulate('click');
        expect(props.signout).toHaveBeenCalled();
    });

    test('Sign up navigation should not be displayed', () => {
        const wrapper = setup(props);
        const signupNavigation = findByAttr(wrapper, 'sign-navigation');
        expect(signupNavigation.length).not.toBe(1);
    });

});

describe('If a user is login with EMPLOYEE role', () => {
    const props = {
        user: {
            isAuthenticated: true,
            data: {
                firstName: 'Admin',
                lastName: 'User',
                role: 'EMPLOYEE',
            }
        },
        signout: jest.fn(),
        history: {
            push: jest.fn()
        }
    };

    test('Admin navaigation section should not be displayed', () => {
        const wrapper = setup(props);
        const adminNavigationSection = findByAttr(wrapper, 'admin-navigation');
        expect(adminNavigationSection.length).not.toBe(1);
    });
});


describe('If a user is not login', () => {
    const props = {
        user: {
            isAuthenticated: false,
            data: null,
        },
        signout: jest.fn(),
        history: {
            push: jest.fn()
        }
    };

    test('Sign up navigation should be displayed', () => {
        const wrapper = setup(props);
        const signupNavigation = findByAttr(wrapper, 'sign-navigation');
        expect(signupNavigation.length).toBe(1);
    });

    test('Check user details section should not be displayed', () => {
        const wrapper = setup(props);
        const userDetailsSection = findByAttr(wrapper, 'user-details-section');
        expect(userDetailsSection.length).not.toBe(1);
    });

});