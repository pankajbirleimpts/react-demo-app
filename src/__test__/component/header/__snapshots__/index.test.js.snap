// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`If a user is login with admin Snapshot of Header component 1`] = `
<Grid
  className="header-container"
>
  <GridRow
    className="header-name-section"
  >
    <GridColumn
      width={12}
    >
      <h3
        className="web-logo"
      >
        Canteen Management System
      </h3>
    </GridColumn>
    <GridColumn
      data-test="user-details-section"
      width={4}
    >
      <div
        className="user-details"
      >
        <p
          className="user-name"
        >
          Admin
           
          User
        </p>
        <p
          className="user-role"
        >
          (
          ADMIN
          )
        </p>
        <a
          data-test="logout-action"
          href="javasctipt:void(0)"
          onClick={[Function]}
        >
          Logout
        </a>
      </div>
    </GridColumn>
  </GridRow>
  <GridRow
    className="menu-container"
  >
    <GridColumn>
      <div
        className="ui secondary pointing menu"
      >
        <MenuLink
          label="Dashboard"
          to="/dashboard"
        />
        <MenuLink
          data-test="admin-navigation"
          label="Users"
          to="/users"
        />
        <MenuLink
          label="Items"
          to="/items"
        />
        <MenuLink
          label="Day Items"
          to="/day-items"
        />
        <MenuLink
          label="Purchase Item"
          to="/purchase-item"
        />
        <MenuLink
          label="Transactions"
          to="/transactions"
        />
      </div>
    </GridColumn>
  </GridRow>
</Grid>
`;
